Add new guide in C4R for Query parameters (#452)
CartoDB/documentation: app/content/react/guides/query-parameters.md

## Query Parameters

> You can use QueryParameters to filter your layers at backend side, you can check our [documentation]({{< ref "/deck-gl/reference#queryparameters-depends-on-provider-optional" >}} "About Us") for different providers.

Reviewer: ```suggestion
You can use QueryParameters to filter your layers at backend side, using the concept of *parametrized queries*. You can check our related [deck.gl documentation]({{< ref "/deck-gl/reference#queryparameters-depends-on-provider-optional" >}} "About Us") for different providers.
```

In this guide we're going to create a simple map that shows the earthquakes in Spain and allows to filter it by their magnitude.

#### Creating a map using QueryParameters

First you'll need to create a source that supports query parameters. In this case we are getting all the earthquakes whoses magnitude are between `min: 4` and `max: 5`.

```ts
import { MAP_TYPES } from '@deck.gl/carto';

> const sql = 'select * from `carto-demo-data.demo_tables.spain_earthquakes` where magnitude @min and @max';

Reviewer: This query lacks BETWEEN, right?

const earthquakesSource = {
  id: 'earthquakes',
  data: sql,
  type: MAP_TYPES.QUERY,
  queryParameters: {
    min: 4,
    max: 5
  }
}
```

> Once you have your source is ready to use you can create your layer using `useCartoLayerProps` hook.

Reviewer: ```suggestion
Once you have your source ready to be used, you can create your layer using `useCartoLayerProps` hook.
```

```ts
import { CartoLayer } from '@deck.gl/carto';
import { useCartoLayerProps } from '@carto/react-api';

const layerProps = useCartoLayerProps({ source: earthquakesSource });
const earthquakesLayer = new CartoLayer({
  ...layerProps,
  id: 'earthquakesLayer',
  pointRadiusUnits: 'pixels',
  lineWidthUnits: 'pixels',
  getLineColor: [0, 0, 0],
  getPointRadius: 3,
});
```

> Your layer is ready and you can use it your map, now we're going to create a `<FormulaWidget />` wich display the number of earthquakes in your viewport.

Reviewer: ```suggestion
Now your layer is ready and you can use it in your map, so we're going to create a `<FormulaWidget />` which displays the number of earthquakes in your viewport.
```

#### Using queryParameters with widgets

> All our widgets are ready to use queryParameters. You just need to pass the datasource properly configured.

Reviewer: ```suggestion
All our widgets are ready to use queryParameters. You just need to pass the `datasource` properly configured (that basically means declaring its `type` as MAP_TYPES.QUERY and adding the corresponding `queryParameters` configuration). After that, you can use any widget as usual, eg:

```ts
<FormulaWidget
  id='totalEarthquakes'
  title='Total earthquakes'
  dataSource={earthquakesSource.id}
  operation={AggregationTypes.COUNT}
  />
```

#### Filtering with queryParameters using a widget

> We can use widgets to modify and apply filters using queryParameters. We're going to add a `<RangeWidgetUI />` to allows filtering by magnitude. First we need to modify the source to support modify `queryParameters`.

Reviewer: ```suggestion
But the utility of queryParameters is dynamically changing the parametrized values in the app, so now we're gonna use a widget to modify those values (and as a consequence, to apply new backend filters using queryParameters). Here we're going to add a `<RangeWidgetUI />` to allow filtering earthquakes by magnitude. 

First we need to prepare the source to support `queryParameters` modification, passing those values as parameters to it.
```

## Query Parameters

> You can use QueryParameters to filter your layers & widgets at the backend side, using the concept of *parametrized queries* over your sources. You can check our related [deck.gl documentation]({{< ref "/deck-gl/reference#queryparameters-depends-on-provider-optional" >}} "About Us") for different providers.

Reviewer: @Josmorsot This "About Us" looks wrong. Maybe a copy & paste. If we keep it, we should display something like "Query parameters in deck.gl"

> You can use QueryParameters to filter your layers & widgets at the backend side, using the concept of *parametrized queries* over your sources. You can check our related [deck.gl documentation]({{< ref "/deck-gl/reference#queryparameters-depends-on-provider-optional" >}} "About Us") for different providers.

Reviewer: ```suggestion
You can use QueryParameters to filter your layers & widgets at the backend side, using the concept of *parametrized queries* over your sources. You can check our related [deck.gl documentation]({{< ref "/deck-gl/reference#queryparameters-depends-on-provider-optional" >}} "queryParameters in CartoLayer") for different providers.
```

In this guide we're going to create a simple map that shows the earthquakes in Spain and allows to filter it by their magnitude.

#### Creating a map using QueryParameters

First you'll need to create a source that supports query parameters. In this case we are getting all the earthquakes whoses magnitude are between `min: 4` and `max: 5`.

```ts
import { MAP_TYPES } from '@deck.gl/carto';

> const sql = 'select * from `carto-demo-data.demo_tables.spain_earthquakes` where magnitude between @min and @max';

Reviewer 2: This syntax (parametrized queries) is only compatible with BigQuery. I think that we should specify it. What about mentioning how would be the same query with another provider

Author: it's just an example. I've added a link to the documentation for queryParameters, where different providers are specified. Anyways I think that would be great to specify that this example is based on BQ

Reviewer 3: I agree with @Josmorsot it'd be enough to point that this example is for BQ and then link to the `queryParameters` documentation with the specifics for each provider.

const earthquakesSource = {
  id: 'earthquakes',
  data: sql,
  type: MAP_TYPES.QUERY,
  queryParameters: {
    min: 4,
    max: 5
  }
}
```

Once you have your source ready to be used, you can create your layer using `useCartoLayerProps` hook.  

```ts
import { CartoLayer } from '@deck.gl/carto';
import { useCartoLayerProps } from '@carto/react-api';

const layerProps = useCartoLayerProps({ source: earthquakesSource });
const earthquakesLayer = new CartoLayer({
  ...layerProps,
  id: 'earthquakesLayer',
  pointRadiusUnits: 'pixels',
  lineWidthUnits: 'pixels',
  getLineColor: [0, 0, 0],
  getPointRadius: 3,
});
```

Now your layer is ready and you can use it in your map, so we're going to create a `<FormulaWidget />` which displays the number of earthquakes in your viewport.

#### Using queryParameters with widgets

> All our widgets are ready to use queryParameters. You just need to pass the `datasource` properly configured (that basically means declaring its `type` as MAP_TYPES.QUERY and adding the corresponding `queryParameters` configuration). After that, you can use any widget as usual, eg:

Reviewer 2: This sentence `All our widgets are ready to use queryParameters` sounds confusing to me. Probably I would say
All our widgets are ready to use **datasources** with queryParameters

Author: Much better
